
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Manipulate the Hierarchy of an AdataDict" />
<meta property="og:type" content="website" />
<meta property="og:url" content="tutorials/adata_dict/Manipulate_the_Hierarchy.html" />
<meta property="og:site_name" content="AnnDictionary" />
<meta property="og:description" content="This tutorial demonstrates how to manipulate the hierarchy of an AdataDict. There are several operations pre-defined. These include: Set the hierarchy, Add a stratification, Flatten the hierarchy. ..." />
<meta property="og:image:width" content="1146" />
<meta property="og:image:height" content="600" />
<meta property="og:image" content="/None" />
<meta property="og:image:alt" content="This tutorial demonstrates how to manipulate the hierarchy of an AdataDict. There are several operations pre-defined. These include: Set the hierarchy, Add a..." />
<meta name="description" content="This tutorial demonstrates how to manipulate the hierarchy of an AdataDict. There are several operations pre-defined. These include: Set the hierarchy, Add a stratification, Flatten the hierarchy. ..." />
<meta name="twitter:card" content="summary_large_image" />

    <title>Manipulate the Hierarchy of an AdataDict &#8212; AnnDictionary Documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/plot_directive.css?v=7f9a90b1" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../_static/documentation_options.js?v=4d2f5802"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'tutorials/adata_dict/Manipulate_the_Hierarchy';</script>
    <link rel="canonical" href="https://ggit12.github.io/anndictionary/tutorials/adata_dict/Manipulate_the_Hierarchy.html" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Iterate Over an AdataDict" href="Iterate_Over_an_AdataDict.html" />
    <link rel="prev" title="Build an AdataDict" href="Build_AdataDict.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">AnnDictionary Documentation</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>

<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/index.html">API</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/adata_dict/index.html">Core</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/adata_dict/adata_dict.html">AdataDict Class</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.set_hierarchy.html">anndict.AdataDict.set_hierarchy</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.add_stratification.html">anndict.AdataDict.add_stratification</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.flatten.html">anndict.AdataDict.flatten</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.fapply.html">anndict.AdataDict.fapply</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.fapply_return.html">anndict.AdataDict.fapply_return</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.index_bool.html">anndict.AdataDict.index_bool</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.set_obs_index.html">anndict.AdataDict.set_obs_index</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.set_var_index.html">anndict.AdataDict.set_var_index</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.check_structure.html">anndict.AdataDict.check_structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.AdataDict.copy.html">anndict.AdataDict.copy</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/adata_dict/adata_dict_fapply.html">Iterate over AnnDictionaries</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.adata_dict.adata_dict_fapply.html">anndict.adata_dict.adata_dict_fapply</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/adata_dict/build_write_read.html">Build, Write, and Read AnnDictionaries</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.adata_dict.build_adata_dict.html">anndict.adata_dict.build_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.adata_dict.add_stratification.html">anndict.adata_dict.add_stratification</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.adata_dict.write_adata_dict.html">anndict.adata_dict.write_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.adata_dict.read_adata_dict.html">anndict.adata_dict.read_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.adata_dict.read_adata_dict_from_h5ad.html">anndict.adata_dict.read_adata_dict_from_h5ad</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/adata_dict/generated/anndict.adata_dict.concatenate_adata_dict.html">anndict.adata_dict.concatenate_adata_dict</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/llm/index.html">LLM Management</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/llm/llm_call.html">Configure and Call</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/llm/generated/anndict.configure_llm_backend.html">anndict.configure_llm_backend</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/llm/generated/anndict.get_llm_config.html">anndict.get_llm_config</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/llm/generated/anndict.call_llm.html">anndict.call_llm</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/llm/generated/anndict.retry_call_llm.html">anndict.retry_call_llm</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/annotate/index.html">Annotate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/annotate/cells/index.html">Cells</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/annotate/cells/de_novo.html">De Novo</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/annotate/cells/label_transfer.html">Transfer Labels</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/annotate/cells/benchmarking.html">Benchmark</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/annotate/cells/error_correction.html">Error Correction</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/annotate/genes/index.html">Genes</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/annotate/genes/make_gene_lists.html">Make Gene Lists</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/plot/index.html">Plot</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/plot/module_score_plots.html">Module Scores</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.module_score_barplot.html">anndict.plot.module_score_barplot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.module_score_umap.html">anndict.plot.module_score_umap</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/plot/annotate_genes_on_heatmap.html">Annotate Genes on Heatmap</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.annotate_gene_groups_with_ai_biological_process.html">anndict.plot.annotate_gene_groups_with_ai_biological_process</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/plot/confusion_matrix_plots.html">Confusion Matrix</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.plot_confusion_matrix_from_adata.html">anndict.plot.plot_confusion_matrix_from_adata</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.plot_confusion_matrix.html">anndict.plot.plot_confusion_matrix</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/plot/sankey_plots.html">Sankey Plots</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.plot_sankey.html">anndict.plot.plot_sankey</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.save_sankey.html">anndict.plot.save_sankey</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/plot/agreement_bar_plots.html">Agreement Bar Plots</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.plot_grouped_average.html">anndict.plot.plot_grouped_average</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.plot_model_agreement.html">anndict.plot.plot_model_agreement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.plot_model_agreement_categorical.html">anndict.plot.plot_model_agreement_categorical</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/plot/stabilizing_classifier_plots.html">Stabilizing Classifier</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.plot_training_history.html">anndict.plot.plot_training_history</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/plot/generated/anndict.plot.plot_label_changes.html">anndict.plot.plot_label_changes</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/automated_label_management/index.html">Automated Label Management</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/automated_label_management/clean_single_column.html">Single Column</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/automated_label_management/generated/anndict.simplify_obs_column.html">anndict.simplify_obs_column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/automated_label_management/generated/anndict.create_label_hierarchy.html">anndict.create_label_hierarchy</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/automated_label_management/generated/anndict.map_cell_type_labels_to_simplified_set.html">anndict.map_cell_type_labels_to_simplified_set</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/automated_label_management/generated/anndict.simplify_var_index.html">anndict.simplify_var_index</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/automated_label_management/generated/anndict.map_gene_labels_to_simplified_set.html">anndict.map_gene_labels_to_simplified_set</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/automated_label_management/unify_columns_within_adata.html">Within Adata</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/automated_label_management/generated/anndict.ensure_label_consistency_adata.html">anndict.ensure_label_consistency_adata</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/automated_label_management/generated/anndict.ensure_label_consistency_main.html">anndict.ensure_label_consistency_main</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/automated_label_management/unify_columns_between_adata.html">Between Adata</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/automated_label_management/generated/anndict.ai_unify_labels.html">anndict.ai_unify_labels</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/utils/index.html">Utils</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/utils/anndictionary.html">AnnDictionary</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.enforce_semantic_list.html">anndict.utils.enforce_semantic_list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.make_names.html">anndict.utils.make_names</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.normalize_string.html">anndict.utils.normalize_string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.normalize_label.html">anndict.utils.normalize_label</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.summarize_metadata.html">anndict.utils.summarize_metadata</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.display_html_summary.html">anndict.utils.display_html_summary</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.create_color_map.html">anndict.utils.create_color_map</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.get_slurm_cores.html">anndict.utils.get_slurm_cores</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/utils/anndata.html">AnnData</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.get_adata_columns.html">anndict.utils.get_adata_columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.remove_genes.html">anndict.utils.remove_genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.add_col_to_adata_obs.html">anndict.utils.add_col_to_adata_obs</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.add_col_to_adata_var.html">anndict.utils.add_col_to_adata_var</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.add_col_to_pd_df.html">anndict.utils.add_col_to_pd_df</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.convert_obs_col_to_category.html">anndict.utils.convert_obs_col_to_category</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.convert_obs_col_to_string.html">anndict.utils.convert_obs_col_to_string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.convert_obs_index_to_str.html">anndict.utils.convert_obs_index_to_str</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/utils/scanpy.html">Scanpy</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.sample_and_drop.html">anndict.utils.sample_and_drop</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.sample_adata_dict.html">anndict.utils.sample_adata_dict</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/utils/pca_density_filter.html">PCA Density Filter</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.pca_density_filter_adata.html">anndict.utils.pca_density_filter_adata</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.pca_density_filter_main.html">anndict.utils.pca_density_filter_main</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/utils/stabilizing_classifier.html">Stabilizing Classifier</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.stable_label_adata.html">anndict.utils.stable_label_adata</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.stable_label.html">anndict.utils.stable_label</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/utils/read_spatial_data.html">Read Spatial Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.build_adata_from_transcript_positions.html">anndict.utils.build_adata_from_transcript_positions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.build_adata_from_visium.html">anndict.utils.build_adata_from_visium</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.add_blank_image_to_adata.html">anndict.utils.add_blank_image_to_adata</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.read_transcript_coords.html">anndict.utils.read_transcript_coords</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.get_steps_and_coords.html">anndict.utils.get_steps_and_coords</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.populate_sparse_array.html">anndict.utils.populate_sparse_array</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.process_gene_counts.html">anndict.utils.process_gene_counts</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.create_anndata.html">anndict.utils.create_anndata</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/utils/uce.html">UCE</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/utils/generated/anndict.utils.uce_adata.html">anndict.utils.uce_adata</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/wrappers/index.html">Fapply Wrappers</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/wrappers/anndictionary.html">AnnDictionary</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.remove_genes_adata_dict.html">anndict.wrappers.remove_genes_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.ai_annotate_biological_process_adata_dict.html">anndict.wrappers.ai_annotate_biological_process_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.simplify_var_index_adata_dict.html">anndict.wrappers.simplify_var_index_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.ensure_label_consistency_adata_dict.html">anndict.wrappers.ensure_label_consistency_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.simplify_obs_column_adata_dict.html">anndict.wrappers.simplify_obs_column_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.create_label_hierarchy_adata_dict.html">anndict.wrappers.create_label_hierarchy_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.ai_annotate_cell_type_adata_dict.html">anndict.wrappers.ai_annotate_cell_type_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.ai_compare_cell_type_labels_pairwise_adata_dict.html">anndict.wrappers.ai_compare_cell_type_labels_pairwise_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.ai_annotate_cell_sub_type_adata_dict.html">anndict.wrappers.ai_annotate_cell_sub_type_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.ai_determine_leiden_resolution_adata_dict.html">anndict.wrappers.ai_determine_leiden_resolution_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.harmony_label_transfer_adata_dict.html">anndict.wrappers.harmony_label_transfer_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.plot_sankey_adata_dict.html">anndict.wrappers.plot_sankey_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.save_sankey_adata_dict.html">anndict.wrappers.save_sankey_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.plot_grouped_average_adata_dict.html">anndict.wrappers.plot_grouped_average_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.plot_label_changes_adata_dict.html">anndict.wrappers.plot_label_changes_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.plot_confusion_matrix_adata_dict.html">anndict.wrappers.plot_confusion_matrix_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.summarize_metadata_adata_dict.html">anndict.wrappers.summarize_metadata_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.display_html_summary_adata_dict.html">anndict.wrappers.display_html_summary_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.pca_density_adata_dict.html">anndict.wrappers.pca_density_adata_dict</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../api/wrappers/scanpy.html">Scanpy</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.sample_and_drop_adata_dict.html">anndict.wrappers.sample_and_drop_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.normalize_adata_dict.html">anndict.wrappers.normalize_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.log_transform_adata_dict.html">anndict.wrappers.log_transform_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.set_high_variance_genes_adata_dict.html">anndict.wrappers.set_high_variance_genes_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.rank_genes_groups_adata_dict.html">anndict.wrappers.rank_genes_groups_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.scale_adata_dict.html">anndict.wrappers.scale_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.pca_adata_dict.html">anndict.wrappers.pca_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.neighbors_adata_dict.html">anndict.wrappers.neighbors_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.leiden_adata_dict.html">anndict.wrappers.leiden_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.leiden_sub_cluster.html">anndict.wrappers.leiden_sub_cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.leiden_sub_cluster_adata_dict.html">anndict.wrappers.leiden_sub_cluster_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.calculate_umap_adata_dict.html">anndict.wrappers.calculate_umap_adata_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../api/wrappers/generated/anndict.wrappers.plot_umap_adata_dict.html">anndict.wrappers.plot_umap_adata_dict</a></li>
</ul>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/wrappers/squidpy.html">Squidpy</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../index.html">Tutorials</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active has-children"><a class="reference internal" href="index.html">Using the AdataDict Class</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Build_AdataDict.html">Build an AdataDict</a></li>
<li class="toctree-l3 current active"><a class="current reference internal" href="#">Manipulate the Hierarchy of an AdataDict</a></li>

<li class="toctree-l3"><a class="reference internal" href="Iterate_Over_an_AdataDict.html">Iterate Over an AdataDict</a></li>
<li class="toctree-l3"><a class="reference internal" href="Write_and_Read_an_AdataDict.html">Write and Read AdataDict</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../annotate/index.html">Annotating with AnnDictionary</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../annotate/automated_cell_type_marker_gene_scores.html">Automated Cell Type Marker Gene Scores</a></li>
<li class="toctree-l3"><a class="reference internal" href="../annotate/cell_type_annotation_with_an_LLM.html">Cell Type Annotation with Large Language Models (LLMs)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../annotate/cell_type_annotation_with_a_label_transfer_model.html">Transfer Labels Using an Sklearn Classifier</a></li>
<li class="toctree-l3"><a class="reference internal" href="../annotate/automated_spatial_transcriptomic_annotation.html">Automated Annotation of Spatial Transcriptomics</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../contribute/index.html">Contribute</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/ggit12/anndictionary" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/tutorials/adata_dict/Manipulate_the_Hierarchy.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Manipulate the Hierarchy of an AdataDict</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Manipulate the Hierarchy of an AdataDict</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#set-the-hierarchy">Set the hierarchy</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#let-s-say-we-want-the-top-level-to-be-cell-type-and-the-second-level-to-be-phase">Let’s say we want the top level to be cell_type and the second level to be phase</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#add-a-stratification">Add a stratification</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#flatten-the-hiearchy">Flatten the hiearchy</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="manipulate-the-hierarchy-of-an-adatadict">
<h1>Manipulate the Hierarchy of an AdataDict<a class="headerlink" href="#manipulate-the-hierarchy-of-an-adatadict" title="Link to this heading">#</a></h1>
<p>This tutorial demonstrates how to manipulate the hierarchy of an AdataDict.
<br></p>
<p>There are several operations pre-defined. These include:</p>
<ol class="arabic simple">
<li><p>Set the hierarchy</p></li>
<li><p>Add a stratification</p></li>
<li><p>Flatten the hierarchy</p></li>
</ol>
</section>
<section id="set-the-hierarchy">
<h1>Set the hierarchy<a class="headerlink" href="#set-the-hierarchy" title="Link to this heading">#</a></h1>
<p>An AdataDict is a nested dictionary of anndata.</p>
<p>The <a class="reference external" href="https://ggit12.github.io/anndictionary/api/adata_dict/adata_dict.html#anndict.AdataDict.hierarchy"><code class="docutils literal notranslate"><span class="pre">.hierarchy</span></code></a> attribute describes the current nesting structure of the object.</p>
<p>The <a class="reference external" href="https://ggit12.github.io/anndictionary/api/adata_dict/generated/anndict.AdataDict.set_hierarchy.html#anndict.AdataDict.set_hierarchy"><code class="docutils literal notranslate"><span class="pre">.set_hierarchy()</span></code></a> method manipulates the nesting structure (aka the hierarchy).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">anndict</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">adt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>

<span class="c1"># Load an example dataset from scanpy</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc68k_reduced</span><span class="p">()</span>

<span class="c1"># Rename obs column (for legibility)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;bulk_labels&#39;</span><span class="p">]</span>


<span class="c1"># Build the AdataDict</span>
<span class="n">adata_dict</span> <span class="o">=</span> <span class="n">adt</span><span class="o">.</span><span class="n">build_adata_dict</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">strata_keys</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span> <span class="s1">&#39;phase&#39;</span><span class="p">])</span> <span class="c1"># Use &#39;cell_type&#39; and &#39;phase&#39; because they&#39;re already in the .obs... You could use any categorical columns.</span>
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">adata_dict</span></code> is now a flat dictionary, where each key is a tuple, and each value in this case has a single cell type-phase group.</p>
<p>We can tell the nesting structure is flat by looking at the <code class="docutils literal notranslate"><span class="pre">.hierarchy</span></code> attribute</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="o">.</span><span class="n">hierarchy</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&#39;cell_type&#39;, &#39;phase&#39;)
</pre></div>
</div>
</div>
</div>
<p>and here’s what the actual object looks like:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{(&#39;CD4+/CD25 T Reg&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 42 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD4+/CD25 T Reg&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 10 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD4+/CD25 T Reg&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 16 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD4+/CD45RA+/CD25- Naive T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 4 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD4+/CD45RA+/CD25- Naive T&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 4 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD4+/CD45RO+ Memory&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 9 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD4+/CD45RO+ Memory&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 10 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD8+ Cytotoxic T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 30 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD8+ Cytotoxic T&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD8+ Cytotoxic T&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 23 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 29 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 2 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 12 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD14+ Monocyte&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 109 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD14+ Monocyte&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 20 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD19+ B&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 65 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD19+ B&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 3 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD19+ B&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 27 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD34+&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 6 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD34+&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 7 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD56+ NK&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 11 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD56+ NK&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;CD56+ NK&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 19 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;Dendritic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 196 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;Dendritic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 44 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;}
</pre></div>
</div>
</div>
</div>
<p>We can re-arrange the structure with the <code class="docutils literal notranslate"><span class="pre">.set_hierarchy()</span></code> method. To use this method, pass a list that is nested in a way that reflects your desired hierarchy.
Here are a few examples, and you can try playing around with it yourself too, to see how it works.</p>
<section id="let-s-say-we-want-the-top-level-to-be-cell-type-and-the-second-level-to-be-phase">
<h2>Let’s say we want the top level to be cell_type and the second level to be phase<a class="headerlink" href="#let-s-say-we-want-the-top-level-to-be-cell-type-and-the-second-level-to-be-phase" title="Link to this heading">#</a></h2>
<p>This could be useful for a pipeline that expects the input to be a single cell type, with separate adata split by phase</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="o">.</span><span class="n">set_hierarchy</span><span class="p">([</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;phase&#39;</span><span class="p">]])</span> <span class="c1"># Note the nesting structure of the input list</span>
</pre></div>
</div>
</div>
</div>
<p>And here’s what the adata_dict looks like now:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="o">.</span><span class="n">hierarchy</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&#39;cell_type&#39;, (&#39;phase&#39;,))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{(&#39;CD4+/CD25 T Reg&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 42 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 10 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 16 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;CD4+/CD45RA+/CD25- Naive T&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 4 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 4 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;CD4+/CD45RO+ Memory&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 9 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 10 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;CD8+ Cytotoxic T&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 30 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 1 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 23 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 29 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 2 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 12 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;CD14+ Monocyte&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 109 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 20 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;CD19+ B&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 65 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 3 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 27 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;CD34+&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 6 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 7 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;CD56+ NK&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 11 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 1 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 19 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;},
 (&#39;Dendritic&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 196 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;,
  (&#39;S&#39;,): AnnData object with n_obs × n_vars = 44 × 765
      obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
      var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
      uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
      obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
      varm: &#39;PCs&#39;
      obsp: &#39;distances&#39;, &#39;connectivities&#39;}}
</pre></div>
</div>
</div>
</div>
<p>Note that adata_dict[(‘CD4+/CD25 T Reg’,)] now gives an AdataDict containing all the phases for that cell type (the AdataDict has a separate key-value for each phase).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="p">[(</span><span class="s1">&#39;CD4+/CD25 T Reg&#39;</span><span class="p">,)]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 42 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 10 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;S&#39;,): AnnData object with n_obs × n_vars = 16 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;}
</pre></div>
</div>
</div>
</div>
<p>Note the nested AdataDicts, where only terminal nodes are AnnData:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="o">.</span><span class="vm">__class__</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>anndict.adata_dict.adata_dict.AdataDict
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="p">[(</span><span class="s1">&#39;CD4+/CD25 T Reg&#39;</span><span class="p">,)]</span><span class="o">.</span><span class="vm">__class__</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>anndict.adata_dict.adata_dict.AdataDict
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="p">[(</span><span class="s1">&#39;CD4+/CD25 T Reg&#39;</span><span class="p">,)][(</span><span class="s1">&#39;G1&#39;</span><span class="p">,)]</span><span class="o">.</span><span class="vm">__class__</span> <span class="c1"># Only the terminal nodes are AnnData</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>anndata._core.anndata.AnnData
</pre></div>
</div>
</div>
</div>
<p>While we show the type of each level of the dictionary here, in practice, you can generally ignore this. The main method of an AdataDict (<a class="reference external" href="https://ggit12.github.io/anndictionary/api/adata_dict/generated/anndict.AdataDict.fapply.html#anndict.AdataDict.fapply"><code class="docutils literal notranslate"><span class="pre">.fapply()</span></code></a>, covered by the “Iterate Over AdataDict” tutorial) is meant to abstract this away and handles automatically looping over the nested structure of the AdataDict.</p>
</section>
<section id="add-a-stratification">
<h2>Add a stratification<a class="headerlink" href="#add-a-stratification" title="Link to this heading">#</a></h2>
<p>Now, let’s say we also wanted to group by another variable (like tissue or is_control). Here, we’ll use the louvain column because it’s already in the dataset.</p>
<p>We can add this as a grouping variable with the <a class="reference external" href="https://ggit12.github.io/anndictionary/api/adata_dict/generated/anndict.AdataDict.add_stratification.html#anndict.AdataDict.add_stratification"><code class="docutils literal notranslate"><span class="pre">.add_stratification()</span></code></a> method as follows. This will set ‘louvain’ as the top-level key and nest the existing structure below that.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="o">.</span><span class="n">add_stratification</span><span class="p">([</span><span class="s1">&#39;louvain&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<p>Note the new hierarchy:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="o">.</span><span class="n">hierarchy</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&#39;louvain&#39;, (&#39;cell_type&#39;, (&#39;phase&#39;,)))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{(&#39;0&#39;,): {(&#39;CD4+/CD25 T Reg&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 41 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 12 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 26 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 9 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD8+ Cytotoxic T&#39;,): {(&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 6 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 9 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD4+/CD45RA+/CD25- Naive T&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 2 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 2 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD4+/CD45RO+ Memory&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 8 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 6 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD56+ NK&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD19+ B&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 2 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 2 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;Dendritic&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;3&#39;,): {(&#39;CD4+/CD25 T Reg&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 3 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 2 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD8+ Cytotoxic T&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 16 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 16 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD56+ NK&#39;,): {(&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 11 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 18 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD4+/CD45RO+ Memory&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;9&#39;,): {(&#39;CD4+/CD25 T Reg&#39;,): {(&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 9 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 3 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD4+/CD45RA+/CD25- Naive T&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,): {(&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD4+/CD45RO+ Memory&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 3 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD19+ B&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;8&#39;,): {(&#39;CD4+/CD45RA+/CD25- Naive T&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD19+ B&#39;,): {(&#39;G2M&#39;,): AnnData object with n_obs × n_vars = 3 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 3 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 23 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;4&#39;,): {(&#39;CD4+/CD45RA+/CD25- Naive T&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD8+ Cytotoxic T&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 2 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD19+ B&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 21 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 40 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD14+ Monocyte&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;5&#39;,): {(&#39;CD4+/CD45RO+ Memory&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD14+ Monocyte&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 4 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 17 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;CD8+ Cytotoxic T&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 3 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;Dendritic&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 8 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 21 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;10&#39;,): {(&#39;CD34+&#39;,): {(&#39;G1&#39;,): AnnData object with n_obs × n_vars = 6 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;S&#39;,): AnnData object with n_obs × n_vars = 7 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;1&#39;,): {(&#39;CD14+ Monocyte&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 15 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 86 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;Dendritic&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 4 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 18 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;6&#39;,): {(&#39;CD14+ Monocyte&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 1 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 5 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;},
  (&#39;Dendritic&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 3 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 33 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;2&#39;,): {(&#39;Dendritic&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 20 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 97 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}},
 (&#39;7&#39;,): {(&#39;Dendritic&#39;,): {(&#39;S&#39;,): AnnData object with n_obs × n_vars = 8 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;,
   (&#39;G1&#39;,): AnnData object with n_obs × n_vars = 27 × 765
       obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
       var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
       uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
       obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
       varm: &#39;PCs&#39;
       obsp: &#39;distances&#39;, &#39;connectivities&#39;}}}
</pre></div>
</div>
</div>
</div>
<p>Note that this AdataDict has 3 levels of nesting, so to access an actual adata, you’d need to do this:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="p">[(</span><span class="s1">&#39;0&#39;</span><span class="p">,)][(</span><span class="s1">&#39;CD4+/CD25 T Reg&#39;</span><span class="p">,)][(</span><span class="s1">&#39;S&#39;</span><span class="p">,)]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 12 × 765
    obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
    var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
    uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="flatten-the-hiearchy">
<h2>Flatten the hiearchy<a class="headerlink" href="#flatten-the-hiearchy" title="Link to this heading">#</a></h2>
<p>Sometimes it’s useful to flatten the adata_dict (i.e. remove the nesting). This can be done with the <a class="reference external" href="https://ggit12.github.io/anndictionary/api/adata_dict/generated/anndict.AdataDict.flatten.html"><code class="docutils literal notranslate"><span class="pre">.flatten()</span></code></a> method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Note that the hierarchy is now flat:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="o">.</span><span class="n">hierarchy</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&#39;louvain&#39;, &#39;cell_type&#39;, &#39;phase&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{(&#39;0&#39;,
  &#39;CD4+/CD25 T Reg&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 41 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD4+/CD25 T Reg&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD4+/CD25 T Reg&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 12 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 26 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 9 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD8+ Cytotoxic T&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD8+ Cytotoxic T&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 6 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD8+ Cytotoxic T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 9 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD4+/CD45RA+/CD25- Naive T&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 2 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD4+/CD45RA+/CD25- Naive T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 2 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD4+/CD45RO+ Memory&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 8 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD4+/CD45RO+ Memory&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 6 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD56+ NK&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD19+ B&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 2 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;CD19+ B&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 2 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;0&#39;,
  &#39;Dendritic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD4+/CD25 T Reg&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD4+/CD25 T Reg&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 3 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 2 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD8+ Cytotoxic T&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 16 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD8+ Cytotoxic T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 16 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD56+ NK&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD56+ NK&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 11 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD56+ NK&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 18 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;3&#39;,
  &#39;CD4+/CD45RO+ Memory&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;9&#39;,
  &#39;CD4+/CD25 T Reg&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 9 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;9&#39;,
  &#39;CD4+/CD25 T Reg&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 3 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;9&#39;,
  &#39;CD4+/CD45RA+/CD25- Naive T&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;9&#39;,
  &#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;9&#39;,
  &#39;CD8+/CD45RA+ Naive Cytotoxic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;9&#39;,
  &#39;CD4+/CD45RO+ Memory&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 3 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;9&#39;,
  &#39;CD19+ B&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;8&#39;,
  &#39;CD4+/CD45RA+/CD25- Naive T&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;8&#39;,
  &#39;CD4+/CD45RA+/CD25- Naive T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;8&#39;,
  &#39;CD19+ B&#39;,
  &#39;G2M&#39;): AnnData object with n_obs × n_vars = 3 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;8&#39;,
  &#39;CD19+ B&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 3 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;8&#39;,
  &#39;CD19+ B&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 23 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;4&#39;,
  &#39;CD4+/CD45RA+/CD25- Naive T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;4&#39;,
  &#39;CD8+ Cytotoxic T&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;4&#39;,
  &#39;CD8+ Cytotoxic T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 2 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;4&#39;,
  &#39;CD19+ B&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 21 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;4&#39;,
  &#39;CD19+ B&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 40 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;4&#39;,
  &#39;CD14+ Monocyte&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;5&#39;,
  &#39;CD4+/CD45RO+ Memory&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;5&#39;,
  &#39;CD14+ Monocyte&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 4 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;5&#39;,
  &#39;CD14+ Monocyte&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 17 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;5&#39;,
  &#39;CD8+ Cytotoxic T&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 3 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;5&#39;,
  &#39;Dendritic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 8 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;5&#39;,
  &#39;Dendritic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 21 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;10&#39;,
  &#39;CD34+&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 6 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;10&#39;,
  &#39;CD34+&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 7 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;1&#39;,
  &#39;CD14+ Monocyte&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 15 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;1&#39;,
  &#39;CD14+ Monocyte&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 86 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;1&#39;,
  &#39;Dendritic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 4 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;1&#39;,
  &#39;Dendritic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 18 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;6&#39;,
  &#39;CD14+ Monocyte&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 1 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;6&#39;,
  &#39;CD14+ Monocyte&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 5 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;6&#39;,
  &#39;Dendritic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 3 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;6&#39;,
  &#39;Dendritic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 33 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;2&#39;,
  &#39;Dendritic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 20 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;2&#39;,
  &#39;Dendritic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 97 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;7&#39;,
  &#39;Dendritic&#39;,
  &#39;S&#39;): AnnData object with n_obs × n_vars = 8 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;,
 (&#39;7&#39;,
  &#39;Dendritic&#39;,
  &#39;G1&#39;): AnnData object with n_obs × n_vars = 27 × 765
     obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
     var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
     uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
     varm: &#39;PCs&#39;
     obsp: &#39;distances&#39;, &#39;connectivities&#39;}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_dict</span><span class="p">[(</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;CD4+/CD25 T Reg&#39;</span><span class="p">,</span> <span class="s1">&#39;G1&#39;</span><span class="p">)]</span> <span class="c1">#AnnData is indexed by a tuple of strings</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 41 × 765
    obs: &#39;bulk_labels&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;S_score&#39;, &#39;G2M_score&#39;, &#39;phase&#39;, &#39;louvain&#39;, &#39;cell_type&#39;
    var: &#39;n_counts&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable&#39;
    uns: &#39;bulk_labels_colors&#39;, &#39;louvain&#39;, &#39;louvain_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
</section>
</section>


                </article>
              

              
              
              
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Manipulate the Hierarchy of an AdataDict</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#set-the-hierarchy">Set the hierarchy</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#let-s-say-we-want-the-top-level-to-be-cell-type-and-the-second-level-to-be-phase">Let’s say we want the top level to be cell_type and the second level to be phase</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#add-a-stratification">Add a stratification</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#flatten-the-hiearchy">Flatten the hiearchy</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By ggit12
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, ggit12.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>